### ISSUE TITLE
[beta.55] private fields transform produces unnecessary WeakMaps, and not-private output.

### ISSUE THREAD

**Issue by trusktr** [babel/babel#8421]
## Bug Report

I played around with the new private fields [in the repl](https://babeljs.io/repl/build/master/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBiD29oG8BQ1oGIBmjoF5oBWdLAIzACcDoA2UgFwFMIGAKAShVI2HgDsI8EEwB0IeAHM2DABYBLCKJyIANNDmLlFSh1IBfVIdR9BDaNhr8mAdziJU2Uc1acgA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=7.0.0-beta.55&envVersion=) and discovered some things:

- The fields aren't truly private. The output code can be imported by script tag (for example) and it leaks the private fields into public space.
  - They are only "private" strictly to the Babel output code.
  - See this [codepen example](https://codepen.io/trusktr/project/editor/82a736a7537498166dac1ad31a5a554c) of public code accessing private fields.
- Unnecessary number of WeakMaps are created, one for each field.
  - If the fields aren't truly private, there's no need to create `WeakMaps` at all.
- We don't know where users will stick output code (f.e. a `<script>` tag, not necessarily a closure), so output should be truly-private, otherwise the performance hit of making `WeakMap`s is entirely defeated.
  - In many cases people might use Node `require` or `Webpack`, which automatically wraps code in a `function`, but this isn't always the case.

See my truly-private implementation idea: https://github.com/babel/proposals/issues/12#issuecomment-410115104

Please feel free to borrow any ideas from there.

**Comment by ljharb** [comment]
This seems a more general problem, unrelated to private fields - if the file is output as a Script, it should be inside an IIFE - otherwise it will create a bunch of global variables.

**Comment by ljharb** [comment]
However, I also don't think anyone is (or should be) sticking babel output directly in a script tag - a bundler is required for modern web dev, and I don't think we need to optimize for people using last-decade practices.

**Comment by trusktr** [comment]
An ideal world it will be when everyone is using modules! I myself use script tags to prototype things, exposing globals. It's just easy.

**Comment by ljharb** [comment]
"easy" and "safe" (and "easy" and "worth babel worrying about") aren't the same thing :-)

**Comment by loganfsmyth** [comment]
Babel 100% assumes that you're using modules at the moment, both in that it creates top-level declarations that are expected not to leak, and that it often generates variable names that will collide across scripts. I'm all for exploring options for resolving that, I don't think it's particularly tied to private fields.

**Comment by trusktr** [comment]
@loganfsmyth Putting the encapsulation concern aside, there's the other issues (f.e. one `WeakMap` per private prop is excessive).

**Comment by loganfsmyth** [comment]
Is size of output code the primary concern? It seems like the number of WeakMaps is just an implementation detail.

**Comment by ljharb** [comment]
@ishitatsuyuki symbols aren't private, in any way, shape, or form.

**Comment by trusktr** [comment]
> Is size of output code the primary concern? It seems like the number of WeakMaps is just an implementation detail.

No, not code size, just that it is wasteful on CPU/Memory resources. There only needs to be at most one `WeakMap` per class. But if there's multiple classes defined in the same scope (f.e. the same file or closure) then there only needs to be one `WeakMap` in that scope.

Currently there's _one WeakMap per private property_ which is overkill in terms of resource usage.

**Comment by trusktr** [comment]
> @ishitatsuyuki symbols aren't private, in any way, shape, or form.

But they're not much worse than these almost-private WeakMaps. It'd be very difficult to make meaningful use of the Symbol keys from the outside.

**Comment by loganfsmyth** [comment]
> The current is wasteful on resources.

Do you have statistics to back that up? I generally ignore any claims about resources and performance in JS unless you've actually measured and compared on various engines.

I don't really disagree that it's an option, I just disagree that it realistically makes any difference for 99% of cases, especially with no numbers to back up your points.

**Comment by ishitatsuyuki** [comment]
https://jsperf.com/weakmap-vs-property/1

**Comment by loganfsmyth** [comment]
@ishitatsuyuki That's not quite the same. I'm 100% behind a version of private fields that uses name-mangling instead of WeakMaps in order to improve performance by using real props (once we've gotten the WeakMap implementation landed anyway), but @trusktr seems focused on the separation between a single WeakMap vs many, and I'm not as convinced that that one vs many makes much of a difference.

**Comment by nicolo-ribaudo** [comment]
@ishitatsuyuki We have loose mode that does exactly this :slightly_smiling_face:

**Comment by trusktr** [comment]
> once we've gotten the WeakMap implementation landed anyway

Well, good to just get it working first anyway!

> Do you have statistics to back that up?

No runtime tests, but here's some numbers:

**Currently:** 100 classes each with 100 private properties in 100 modules (1 class per file) means 10,000 WeakMaps (because 100 classes x 100 properties, regardless of the number of modules). Each private property of each class instance needs 1 object for the property metadata, so if we have 1 instance of each class, that's 10,000 more objects, for 20,000 total.

**With one-weakmap-per-module:** For the same setup as the previous 100 classes/files, there will be 100 WeakMaps (one per file/module). Each WeakMap will need one cache object per class instance. This will be 100 cache objects if there is 1 instance of each class. In each cache object, there will still need to be the same number of private property metadata objects, which is 10,000. The total is 10,200.

If we disregard the property metadata objects count which remains the same in both methods, we have 10,000 vs 200.

---

If we spread those classes across 50 files instead of 100 (for example), this leaves the current number of WeakMaps the same, but with weakmap-per-module we now have 100 objects instead of 200.

---

For [this input](https://babeljs.io/repl/build/master/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBiD29oG8BQ1oGIBmjoF5oBWdLAIzACcDoA2UgFwFMIGAKAShVI2HgDsI8EEwB0IeAHM2DABYBLCKJyIANNDmLlFSh1IBfVIdR9BDaNhr8mAdziJU2Uc1acgA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=7.0.0-beta.55&envVersion=),

```js
class Foo {
  #foo = 5
  #bar = 6
  test() {
    console.log(this.#foo, this.#bar)
  }
}

const f = new Foo
f.test()
```

the output would be more like

```js
"use strict";

function _instanceof(left, right) { if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) { return right[Symbol.hasInstance](left); } else { return left instanceof right; } }

function _classCallCheck(instance, Constructor) { if (!_instanceof(instance, Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var _privatesMap = new WeakMap // the single WeakMap in the module

function _private(obj) {
  var privates = _privatesMap.get(this);
  if (!privates) _privatesMap.set(this, privates = {});
  return privates;
}

var Foo =
/*#__PURE__*/
function () {
  function Foo() {
    _classCallCheck(this, Foo);

    _private(this).foo = {
      writable: true,
      value: 5
    };

    _private(this).bar = {
      writable: true,
      value: 6
    };
  }

  _createClass(Foo, [{
    key: "test",
    value: function test() {
      console.log(_private(this).foo.value, _private(this).bar.value);
    }
  }]);

  return Foo;
}();
```

**Comment by ljharb** [comment]
How many classes have 100 properties in their entire prototype chain, let alone have 100 own public properties, let alone will have 100 private properties? I'm not sure that's a realistic test case.
